### PR TITLE
Type of Crossing

### PR DESCRIPTION
**Pull request by ENT8R**
Like mentioned in issue #397, this PR adds a quest which will allow the user to set the type of a crossing.
This is my first PR, so I have a few questions:

1. Is the importance (lowest) of the quest right?
2. I used the quest icon "pedestrian_crossing" but this is used by another question too... So do we need a new icon for this quest?
3. I wanted to import some images for each answer but I didn't know how to do this... So I used a hydrant as a placeholder.

But the images I found so far are:

- [Traffic lights](http://maxpixel.freegreatpicture.com/Green-Traffic-Lights-Little-Green-Man-Go-819851)
- [Uncontrolled crossing](https://commons.wikimedia.org/wiki/File%3AZebra_crossing.jpg)
- [Traffic island](https://commons.wikimedia.org/wiki/File%3ATraffic_island%2C_Ballymagorry%2C_May_2010.JPG)

I haven't found an image which could describe `crossing=unmarked`. Maybe somebody has an idea for that?
As I said above, I didn't know how to import those images... Could somebody teach me how to do this in the future?

Thank you for answering my questions! :smiley:

### CODE REVIEW

File: app/src/main/java/de/westnordost/streetcomplete/quests/crossing_type/AddCrossingTypeForm.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+import android.view.View;
+import android.view.ViewGroup;
+
+import de.westnordost.streetcomplete.R;
+import de.westnordost.streetcomplete.quests.ImageListQuestAnswerFragment;
+
+public class AddCrossingTypeForm extends ImageListQuestAnswerFragment
+{
+    private final OsmItem[] TYPES = new OsmItem[] {
+            new OsmItem("traffic_signals", R.drawable.fire_hydrant_pillar, R.string.quest_crossing_type_signals),
```
Comment (krzyk) [inline app/src/main/java/de/westnordost/streetcomplete/quests/crossing_type/AddCrossingTypeForm.java]: you probably want a different drawing for crossings, right now you use fire hydrant :)

File: app/src/main/java/de/westnordost/streetcomplete/quests/crossing_type/AddCrossingTypeForm.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+    @Override protected int getItemsPerRow()
+    {
+        return 4;
+    }
+
+    @Override protected int getMaxNumberOfInitiallyShownItems()
+    {
+        return 4;
+    }
```
Comment (westnordost) [inline app/src/main/java/de/westnordost/streetcomplete/quests/crossing_type/AddCrossingTypeForm.java]: You can leave this out.

File: app/src/main/java/de/westnordost/streetcomplete/quests/crossing_type/AddCrossingTypeForm.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+import de.westnordost.streetcomplete.R;
+import de.westnordost.streetcomplete.quests.ImageListQuestAnswerFragment;
+
+public class AddCrossingTypeForm extends ImageListQuestAnswerFragment
+{
+    private final OsmItem[] TYPES = new OsmItem[] {
+            new OsmItem("traffic_signals", R.drawable.fire_hydrant_pillar, R.string.quest_crossing_type_signals),
+            new OsmItem("uncontrolled", R.drawable.fire_hydrant_pillar, R.string.quest_crossing_type_uncontrolled),
+            new OsmItem("island", R.drawable.fire_hydrant_pillar, R.string.quest_crossing_type_island),
+            new OsmItem("unmarked", R.drawable.fire_hydrant_pillar, R.string.quest_crossing_type_unmarked)
```
Comment (westnordost) [inline app/src/main/java/de/westnordost/streetcomplete/quests/crossing_type/AddCrossingTypeForm.java]: Missing images. Perhaps icons would be better than photos for this?

Comment (ENT8R) [inline app/src/main/java/de/westnordost/streetcomplete/quests/crossing_type/AddCrossingTypeForm.java]: I can't find icons that could describe those answers... I would prefer images, but I don't know how to import them. I tried the Android Studio Image Asset Generator but that doesn't work...

**Comment by Etua** [comment]
Is it possible to mark multiple types in this quest for example: island + traffic lights?

**Comment by ENT8R** [comment]
No this is not possible.
But do you think this would make sense? If yes, somebody has to write a new kind of QuestAnswerFragment because at the moment you can only choose one value per quest...

**Comment by Etua** [comment]
Yes, that makes sense (if correct for OSM editing standards) as there are some crossings with both island and traffic lights that may even work in different scheme depending on the side of the island.

**Comment by Etua** [comment]
Yes, that makes sense (if correct for OSM editing standards) as there are some crossings with both island and traffic lights that may even work in different scheme depending on the side of the island.

**Comment by ENT8R** [comment]
The [wiki](http://wiki.openstreetmap.org/wiki/Multiple_values) says that it is allowed to have multiple values. But I think the problem is, that StreetComplete doesn't offer the possibility to make a quest with more than one answer. You will have to write a new QuestAnswerFragment for that...

**Comment by westnordost** [comment]
> StreetComplete doesn't offer the possibility to make a quest with more than one answer

Yes It does. Even specifically the ImageSelectAdapter. The sports quest uses that functionality.

> I didn't know how to import those images... Could somebody teach me how to do this in the future?

There is no magic there, just normal Android resource management. If you need a refresher, see the [api guide](https://developer.android.com/guide/topics/resources/providing-resources.html) or just look how it has been done for the other assets in StreetComplete.

> Use of images in general

You should search for images / icons that are generic enough that they are understood all over the world, that is why I proposed using icons. These could be mashed up from the already existing icons for traffic lights, zebra crossing and roads.
And generally of course, it is worth a consideration whether having images that accompany the name / description actually adds any value to it. Does it aid the user to select the right choice? If not, a pure radio group / checkbox group
```
( ) option 1          ( ) option 2
( ) option 3          ( ) option 4
```
will serve the same purpose and is a little smaller. I am not implying that images make no sense here, just saying in general because with this quest type I am not sure if the text wouldn't be enough.

**Comment by ENT8R** [comment]
Ooops...
Then I will update the code with the functionality to select multiple items and answers which consist of text only

**Comment by westnordost** [comment]
Well, as said, I'd say it's open for discussion. What do you think about it?

**Comment by ENT8R** [comment]
I updated the code, so the user can select multiple items. And I added images for every answer, because I think in this case they are the best choice...

File: app/src/main/java/de/westnordost/streetcomplete/quests/crossing_type/AddCrossingTypeForm.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    }
+
+    @Override protected OsmItem[] getItems()
+    {
+        return TYPES;
+    }
+
+    @Override protected int getItemsPerRow()
+    {
+        return 4;
```
Comment (westnordost) [inline app/src/main/java/de/westnordost/streetcomplete/quests/crossing_type/AddCrossingTypeForm.java]: I am pretty sure that 4 items in a row but texts like "Crossing with traffic-island" will not work, this is too crammed. As (IMO) the picture is of secondary importance, my suggestion is to use a different layout for the image selector here:

`imageSelector.setCellLayout(R.layout.small_image_with_label_at_end);`

... which is like this:
```
[ image ] Crossing with traffic lights
[ image ] Crossing with traffic-island
```
(You need to create this layout, it does not exist yet. The image would be smaller then, so that it doesn't look odd with the one line of text.)

What do you think?

File: app/src/main/java/de/westnordost/streetcomplete/quests/crossing_type/AddCrossingTypeForm.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    private final OsmItem[] TYPES = new OsmItem[] {
+            new OsmItem("traffic_signals", R.drawable.crossing_type_signals, R.string.quest_crossing_type_signals),
+            new OsmItem("uncontrolled", R.drawable.crossing_type_zebra, R.string.quest_crossing_type_uncontrolled),
+            new OsmItem("island", R.drawable.crossing_type_island, R.string.quest_crossing_type_island),
+            new OsmItem("unmarked", R.drawable.crossing_type_unmarked, R.string.quest_crossing_type_unmarked)
+    };
+
+    @Override protected int getMaxSelectableItems()
+    {
+        return -1;
```
Comment (westnordost) [inline app/src/main/java/de/westnordost/streetcomplete/quests/crossing_type/AddCrossingTypeForm.java]: This allows the user to input invalid data, as traffic signals, uncontrolled and unmarked are mutually exclusive. This is not good, because it can be avoided.

The `island` tag and in particular that it can be combined with any other of these values but the other values cannot be combined with each other appears somewhat problematic to me. This calls for ("schreit nach") `island` should really be a separate tag. [It has also been noted on the discussion page](https://wiki.openstreetmap.org/wiki/Talk:Key:crossing#Island_.2Aor.2A_Traffic_signals_.3F). Also note that there is a big warning at the start of the page _"This page is being considered for cleanup. Please discuss this page."_, so I'd say at least there is agreement on that the wiki page on the crossing tag is problematic.

0.25% of all `crossing=`-tags use multiple values (with `island`). This is really insignificant.

In my opinion, there are two ways to solve this:
1. Since 0.25% is _really_ insignificant, and given that that tagging scheme is problematic and discussed anyway, simply do not implement this multi-choice. `island` then is to be interpreted as "no traffic lights, no markings, but... an island". This is how I imagine this tag was originally conceived for before other osm-wiki-people found that it could be combined with other values as well.
2. Do not show this option at all because it is not clear anyway what a simple crossing=island tag would imply (Does it imply `crossing=unmarked`? Not defined). Unlike the other three values (uncontrolled/unmarked/with traffic lights ) It is of no use for routing applications.
Then, maaaybe there can be a second quest which asks whether there is an island for this crossing or not which tags `crossing:island=yes/no`. But, to be honest, the value of specifying whether there is an island or not is completely oblivious to me. (And also `crossing:island` is undocumented)

File: app/src/main/java/de/westnordost/streetcomplete/quests/crossing_type/AddCrossingTypeForm.java (hunk 4 above)
Comment (ENT8R) [inline app/src/main/java/de/westnordost/streetcomplete/quests/crossing_type/AddCrossingTypeForm.java]: Do you like that?

**Comment by ENT8R** [comment]
Now I removed the the option to select multiple items, added a new layout ([image](https://user-images.githubusercontent.com/25306497/28823684-20b78b64-76bf-11e7-96f5-e1fd38af7b0a.png)) and removed the traffic island question. I think it is OK that the user tags a crossing with a traffic island as `unmarked` because, like you said, it is of no use for routing applications and actually a traffic island is some kind of unmarked crossing...

**Comment by ENT8R** [comment]
While testing the new layout, I discovered an issue: The answer is only selected if I click on the image. If I click on the text, nothing happens... But because the image is so small in my new layout, it would make sense that if the user clicks on the text, the answer is selected too... How can I fix this?

**Comment by westnordost** [comment]
> But because the image is so small in my new layout [...]
Small? It's huge! You know, I suggested changing the layout because I was worried about that it takes up too much vertical space :-D. Now that I see it in action and that there are only 3 options left, perhaps that suggestion isn't that great after all since the texts could also just be shortened to "with traffic lights", "uncontrolled", "no road markings" and thus probably fit in 1-2 lines of text.

> [...] it would make sense that if the user clicks on the text, the answer is selected too... [...]
Hmm, this is odd, the `ImageSelectAdapter `registers a `ClickListener `on the whole view (=whole layout) of the item displayed, not only on the image. What if you give the layout a solid background color (in this case, the drawable/image_select_cell, is the click registered then?

**Comment by Etua** [comment]
@ENT8R As you may be right about the lack of purpose of marking trafic island in terms of routing, I cannot agree that an island is always unmarked.

**Comment by ENT8R** [comment]
I will think about the layout once again...

> What if you give the layout a solid background color (in this case, the drawable/image_select_cell, is the click registered then?

No.

**Comment by krzyk** [comment]
@Etua Actually I would argue otherwise, all islands I've seen in Poland are marked, so it is a type of `uncontrolled` crossing. I think that if someone makes an effort to put the island in the middle of road, they will also draw some lines on the road.

But for the purpose of crossing, island is a feature of the crossing not a type, because it is possible to have island in all other 3 cases.

**Comment by ENT8R** [comment]
@krzyk But in Germany there are almost only traffic islands which are **NOT** marked...

**Comment by krzyk** [comment]
@ENT8R I don't think it makes sense to add it, because the only sensible solution would be to have additional tag `island=yes/no`.
Because you can have:
- traffic_lights + island
- uncontrolled + island
- unmarked + island

So as @westnordost suggested somewhere in this PR, it would make sense to add it as a separate quest (if we want to push unconfirmed tag, that is).

**Comment by ENT8R** [comment]
Should we show a hint to the user, that the user can't answer the quest with `traffic-island`?

**Comment by Etua** [comment]
@ENT8R Do you mean that there will be a separate quest for that or ttaffic lights + island should not be correct anwser at all?

**Comment by ENT8R** [comment]
For this quest this answer should not be a correct answer at all. And I think a seperate quest does not make sense, because information about traffic island is not important

**Comment by Etua** [comment]
@ENT8R Why do you think that island + traffic lights should not be a correct anwser when there are examples of such combination?

**Comment by ENT8R** [comment]
Because you need a new quest to answer this...

**Comment by westnordost** [comment]
Hey @Etua please do not bother ENT8R with islands. The requirement to remove islands comes from me, so argue with me :-)

Here are my reasons for requiring it to be removed:

https://github.com/westnordost/StreetComplete/pull/432/commits/e3afd40adac718ae3262f5465f396f9c6d17ec2b#r130576651

**Comment by Etua** [comment]
@ENT8R Maybe I haven't seen some important issue in this discourse, but what does prevent the quest to be multicheckbox without the possibility to combine unmarked + traffic lights and lights + no lights?

**Comment by westnordost** [comment]
>> What if you give the layout a solid background color (in this case, the drawable/image_select_cell, is the click registered then?
> No.

Then I am out of clues. Perhaps you could put a breakpoint at the mentioned point and see what type the view is on which the listener was registered. Is it the image view after all? What if there are two images or some other elements (some solid boxes) in the layout, will the click be registered on those images?

**Comment by westnordost** [comment]
@Etua What, you mean like

**What properties does this pedestrian crossing have?**
- [ ] traffic lights
- [ ] marking on the ground
- [ ] an island in the middle

**Comment by Etua** [comment]
@westnordost So simply a lot of problems come from an unclear state of the tag. I'm in favor of using it because if not other option is suggested by the wiki, I think that other tag about crossing type without island tag suggests that there is no island (which considering this 0,25% will be a bad assume).

**Comment by Etua** [comment]
@westnordost I think that is the best solution if we want to add island tag. I want to make that clear: does traffic lights imply that the crossing is marked?

**Comment by westnordost** [comment]
It is not defined.

**Comment by westnordost** [comment]
@Etua It is not defined. That is the (or "a") problem.

**Comment by ENT8R** [comment]
What should I do now?
- A quest which leaves `traffic-island` out
- A multi-checkbox quest like @Etua suggested
- Or something completely else?

**Comment by westnordost** [comment]
Leave it out.

**Comment by ENT8R** [comment]
The quest looks now like this:



What do you think?

**Comment by Etua** [comment]
@ENT8R Preety nice.

**Comment by westnordost** [comment]
Let me know when this PR would be ready to merge from your point of view, then I'll review it again.

**Comment by ENT8R** [comment]
I think everything is fine now. I left traffic island and multiple select out. That's it, isn't it?

**Comment by westnordost** [comment]
Ok, I'll look through it later today and merge if all is ok from my side. Thank you for this PR

On 9 August 2017 10:27:24 CEST, ENT8R <notifications@github.com> wrote:
>I think everything is fine now. I left traffic island and multiple
>select out. That's it, isn't it?
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/westnordost/StreetComplete/pull/432#issuecomment-321189210

-- 
Diese Nachricht wurde von meinem Android-Mobiltelefon mit K-9 Mail gesendet.

**Comment by rugk** [comment]
BTW @ENT8R the first image you suggested for traffic lights, is from a site called `http://maxpixel.freegreatpicture.com`. This dubious URL should already give a hint and indeed from the design it really looks like a copycat website of [Pixabay](https://pixabay.com). Do not support such websites! Pixabay even [once switched the license](https://pixabay.com/en/forum/official-pixabay-news-2/a-new-license-for-all-future-pixabay-images-and-vi-4871/) away from CC0 to prevent that, but (thankfully) reverted that decision.
"Thanks" to such website, we however, do not have real CC0 images on [Unsplash](https://community.unsplash.com/articles/unsplashlicense) e.g. 
So TL;DR: Do not link/support such dubious websites. Use Pixabay, Unsplash (even Pexels may be okay – they do things differently, at least, and not just copy that) or other well-known sites. Of course 

BTW for anyone with German knowledge or someone keen to read this by using a translation service, see [this great German article](https://irights.info/artikel/streit-um-freie-bilder-fotodienste-unsplash-und-pixabay-stoeren-sich-an-copycats/28657) explaining the problem. (The dubious website you linked here, is explicitly named.)

**Comment by ENT8R** [comment]
Thank you for the hint! I didn't notice it while I implemented the quest...
Should we change the image **and** the URL or just the URL?

**Comment by rugk** [comment]
The image is the same, so it is enough to change the URL. (although I don't know whether/where it is mentioned in the source)
